### PR TITLE
Spring compression

### PR DESCRIPTION
**Pull request by adrosenbaum**
This PR adds spring compression to cg

## How to prepare for test
- [x] ssh to hasta
- [x] install on stage:
`bash servers/resources/hasta.scilifelab.se/update-cg-stage.sh spring`
- [x] do `us`

**There are two tests on hasta for this PR**

## Test 1: compress and decompress files for a case with fastq files

Compress case fastq files by running:
1. `cg compress fastq --case-id vitalmouse`
Clean case fastq files by running:
1. `cg compress clean fastq --case-id vitalmouse`
Decompress spring files by running
1. `cg decompress spring vitalmouse`
Update housekeeper by running 
1. `cg store fastq vitalmouse`

### Expected test outcome

1. Compress case fastq files
    - [x] Check that the spring archive was created (should be next to fastqs) ✅ 
    - [x] Check that the spring archive metadata file was created (ends with `.json`) ✅ 
1. Clean case fastq files
    - [x] Check that the fastq files for the case was deleted
1. Decompress spring files
    - [x] Check that the spring files were decompressed with matching shasum to original fastq files
    - [x] ~Check that the spring archive metadata file was deleted~
Skärmavbild 2020-06-17 kl  14 27 29

1. Store the decompressed fastq files in housekeeper
    - [x] Check that fastq files are added to HK bundle again
Skärmavbild 2020-06-17 kl  14 31 06

1. Clean again
    - [x] Run `cg compress clean fastq --case-id vitalmouse`
    - [x] Check that no files where deleted
    - [x] Modify date in `metadata.json` to a date older than 3 weeks
    - [x] Clean and check that fastq files are deleted and that HK is updated 
    - [x] Take a screenshot and attach or copy/paste the output.

Skärmavbild 2020-06-29 kl  15 17 46

## Test 2: compress and decompress files for a case with multiple fastq files
**This is one of the individuals in strongbison which we will follow in the tests below:**
Skärmavbild 2020-06-18 kl  22 47 36

Compress case fastq files by running:
1. `cg compress fastq --case-id strongbison`
Clean case fastq files by running:
1. `cg compress clean fastq --case-id strongbison`
Decompress spring files by running
1. `cg decompress spring strongbison`
Update housekeeper by running 
1. `cg store fastq strongbison`

### Expected test outcome

1. Compress case fastq files
    - [X] Check that multiple sbatch files where generated
    
Skärmavbild 2020-06-18 kl  23 05 02

- [x] Check that the multiple spring archives was created (should be next to fastqs) ✅ 
- [x] Check that each spring archive has a metadata file (ends with `.json`) ✅ 
Skärmavbild 2020-06-18 kl  23 28 17

2. Clean case fastq files
- [x] Check that all the fastq files for the case was deleted
Skärmavbild 2020-06-18 kl  23 29 38
- [x] Check that the fastq files are replaced by spring files in HK
Skärmavbild 2020-06-18 kl  23 30 51

3. Decompress spring files
- [x] Check that the spring files were decompressed with matching shasum to original fastq files
Skärmavbild 2020-06-19 kl  07 04 19

4. Store the decompressed fastq files in housekeeper
- [x] Check that all fastq files are added to HK bundle again
Skärmavbild 2020-06-19 kl  07 05 57

5. Clean again
- [x] Run `cg compress clean fastq --case-id strongbison`
- [x] Check that no files where deleted since they are not old enough
Skärmavbild 2020-06-19 kl  07 07 26

**Review:**
- [ ] code approved by
- [x] tests executed by @moonso , @henrikstranneheim 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

**Implementation plan**
- [ ] Install crunchy version 0.5 in P_crunchy

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by adrosenbaum** [comment]
As I understand it the fastq files will be in the sample hk bundles. Is this correct? and if so, can we assume that there always will be exactly two fastq files (one  for each read pair) in each of these bundles? @patrikgrenfeldt @henrikstranneheim

**Comment by adrosenbaum** [comment]
Now the command `cg compress fastq --dry-run -n 5 ` Searches for fastq files in the sample bundles for all cases in the status-db until 5 cases has been compressed.

**Comment by henrikstranneheim** [comment]
> As I understand it the fastq files will be in the sample hk bundles. Is this correct? and if so, can we assume that there always will be exactly two fastq files (one for each read pair) in each of these bundles? @patrikgrenfeldt @henrikstranneheim

To my knowledge we only sequence paired-end. I think it's ok to do the happy path here, as you will skip any fastqs that are single end right?

**Comment by adrosenbaum** [comment]
Yes, single read fastqs, if there are any, will be skipped.

**Comment by adrosenbaum** [comment]
There is still some stuff that I think should be done in this PR
- A command or method that decompresses SPRING files
- A command that removes FASTQ files once they have been compressed
You must also discuss how these procedures will be used in the context of reruns of cases. I think that it would be easiest to add what cases have uncompressed FASTQ files to the status-db.

**Comment by moonso** [comment]
Compression part is ready to be tested. I will continue with the cleaning

**Comment by moonso** [comment]
Ready to be tested @henrikstranneheim

**Comment by henrikstranneheim** [comment]
```
[0|107|5207] [henrik.stranneheim@hasta:/home/proj/stage] [base] 130 $ us
[0|107|5207] [henrik.stranneheim@hasta:/home/proj/stage] [S_main] 4s $ cg compress fastq --case-id vitalmouse
Traceback (most recent call last):
  File "/home/proj/bin/conda/envs/S_main/bin/cg", line 11, in <module>
    sys.exit(base())
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1063, in invoke
    Command.invoke(self, ctx)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/cg/cli/compress/base.py", line 20, in compress
    context.obj["db"] = Store(context.obj.get["database"])
TypeError: 'builtin_function_or_method' object is not subscriptable
```

**Comment by henrikstranneheim** [comment]
@moonso Would you like to have a look

**Comment by moonso** [comment]
Fixed that one @henrikstranneheim

**Comment by henrikstranneheim** [comment]
```
[0|107|5088] [henrik.stranneheim@hasta:/home/proj/stage] [S_main] 1m41s $ cg compress fastq --case-id vitalmouse
Searching for FASTQ files in vitalmouse
/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/query.py:2334: SAWarning: Pathed join target File.version has already been joined to; skipping
  "been joined to; skipping" % prop
Traceback (most recent call last):
  File "/home/proj/bin/conda/envs/S_main/bin/cg", line 11, in <module>
    sys.exit(base())
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/proj/bin/conda/envs/S_main/lib/python3.7/site-packages/cg/cli/compress/fastq.py", line 46, in fastq
    first_fastq
TypeError: is_spring_compression_done() missing 1 required positional argument: 'fastq_second'
```
Another one coming up

**Comment by moonso** [comment]
Wow, ok seems like this needs some testing first

**Comment by moonso** [comment]
Blocked by #643

**Comment by henrikstranneheim** [comment]
**How to test**:
Compress some fastq files by running
- [ ] `cg compress fastq --case-id vitalmouse`

**Expected test outcome**:
- [ ] check that the spring archive was created (should be next to fastqs)
Screenshot 2020-06-08 at 09 40 15

- [ ] check that the spring archive metadata file was created (ends with `.json`)

File: cg/cli/compress/fastq.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        cases = [case_obj]
+
+    conversion_count = 0
+    nr_cases = 0
+    for nr_cases, case in enumerate(cases, 1):
+        case_converted = True
+        if conversion_count >= number_of_conversions:
+            continue
+
+        LOG.info("Searching for FASTQ files in %s", case.internal_id)
```
Comment (henrikstranneheim) [inline cg/cli/compress/fastq.py:38]:
```suggestion
        LOG.info("Searching for FASTQ files in case: %s", case.internal_id)
```

**Comment by henrikstranneheim** [comment]
**How to test**:
Clean case fastq files by running:
1. `cg compress clean fastq --case-id vitalmouse`

**Expected test outcome**:
2. Clean case fastq files ✅ 
- [x] Check that the fastq files for the case was deleted
Screenshot 2020-06-09 at 13 49 07

Screenshot 2020-06-09 at 13 51 10

**Comment by moonso** [comment]
Decompress command ready for test @henrikstranneheim !

**Comment by henrikstranneheim** [comment]
Before decompress:
Screenshot 2020-06-16 at 12 59 40

**Comment by moonso** [comment]
Ok @barrystokman , @patrikgrenfeldt , @henrikstranneheim this beast is ready for review. I'm not jealous, sorry that it became huge but there where not many options. I have and will perform all the manual tests. 

Perhaps something for a rainy midsummer eve 😆 
I will be gone next week (26) but in the week after that (27) I'll be back. Hope that we can merge this and start the cronjob for compressing fastqs during week 27.

File: cg/cli/compress/fastq.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        cases = get_fastq_cases(store, case_id)
+    except CaseNotFoundError:
+        return
+
+    conversion_count = 0
+    for case in cases:
+        # Keeps track on if all samples in a case have been converted
+        case_converted = True
+        if conversion_count >= number_of_conversions:
+            continue
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/cli/compress/fastq.py:37]: Perhaps better with a break

File: cg/meta/compress/compress.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        self.ntasks = 12
+        self.mem = 50
+        self.dry_run = dry_run
+
+    def set_dry_run(self, dry_run: bool):
+        """Update dry run"""
+        self.dry_run = dry_run
+        self.crunchy_api.set_dry_run(dry_run)
+
+    def get_latest_version(self, bundle_name: str) -> hk_models.Version:
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/meta/compress/compress.py:40]: A lot of the code is written by Adam and I did not want to completely change the flow. I agree with you but think it should be changed in another PR

File: cg/meta/compress/compress.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+        """Fetch the latest version of a hk bundle"""
+        last_version = self.hk_api.last_version(bundle_name)
+        if not last_version:
+            LOG.warning("No bundle found for %s in housekeeper", bundle_name)
+            return None
+        LOG.debug("Found version obj for %s: %s", bundle_name, repr(last_version))
+        return last_version
+
+    def get_scout_case(self, case_id: str) -> dict:
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/meta/compress/compress.py:49]: Same

File: cg/meta/compress/compress.py [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+                    LOG.warning("Fastq files already exists in housekeeper")
+                    return False
+
+            LOG.info(
+                "Decompressing %s to FASTQ format for sample %s ", spring_path, sample_id,
+            )
+
+            spring_metadata_path = self.crunchy_api.get_flag_path(spring_path)
+            spring_metadata = self.crunchy_api.get_spring_metadata(spring_metadata_path)
+            for file_info in spring_metadata:
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/meta/compress/compress.py:248]: Yes but I will be careful to make high level changes now when this is more or less ready. That could have affects on many parts of the code including tests

File: cg/meta/compress/files.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    HK_BAM_TAGS,
+    HK_FASTQ_TAGS,
+)
+
+LOG = logging.getLogger(__name__)
+
+# Functions to get common files
+
+
+def get_hk_files_dict(tags: List[str], version_obj: hk_models.Version) -> dict:
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/meta/compress/files.py:24]: I think we should take a wider grip of CG in general for things like this after the summer. We have had some discussions on how to make these more general and reusable. For now I wish to leave it here as an example on how I think it should be done

File: cg/meta/compress/files.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        file_tags = {tag.name for tag in file_obj.tags}
+        if not file_tags.intersection(tags):
+            continue
+        LOG.debug("Found file %s", file_obj)
+        path_obj = Path(file_obj.full_path)
+        hk_files_dict[path_obj.resolve()] = file_obj
+    return hk_files_dict
+
+
+def get_nlinks(file_link: Path) -> int:
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/meta/compress/files.py:46]: Same

File: cg/meta/compress/files.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+    spring_paths = []
+    for file_path in hk_files_dict:
+        if file_path.suffix == ".spring":
+            spring_paths.append(file_path)
+
+    return spring_paths
+
+
+def get_run_name(fastq_path: Path) -> str:
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/meta/compress/files.py:169]: I'm not ready to change this things when it's so near to be put in production. I would have to update multiple tests etc if making this type of change. Hope you understand

File: cg/meta/compress/files.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        fastq_first = {"path": sorted_fastqs[0], "hk_file": hk_files_dict[sorted_fastqs[0]]}
+        fastq_second = {"path": sorted_fastqs[1], "hk_file": hk_files_dict[sorted_fastqs[1]]}
+        fastq_dict[run] = {
+            "fastq_first_file": fastq_first,
+            "fastq_second_file": fastq_second,
+        }
+    return fastq_dict
+
+
+def sort_fastqs(fastq_files: List[Path]) -> Tuple[Path, Path]:
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/meta/compress/files.py:219]: Same argumentation as above

File: tests/apps/crunchy/test_crunchy.py [hunk 10]
```diff
@@ -181,14 +193,236 @@ def test_get_index_path(crunchy_config_dict, crunchy_test_dir):
...
+
+    # THEN result should be True since the files are older than 3 weeks
+    assert result is True
+
+
+def test_get_spring_path_from_fastq():
+    """Test to get a spring path"""
+    # GIVEN a fastq path for a read 1 in pair
+    ind_id = "ind1"
+    fastq = Path("".join([ind_id, FASTQ_FIRST_READ_SUFFIX]))
```
[the remark this answers is not in the document]
Comment (moonso) [reply tests/apps/crunchy/test_crunchy.py:351]: This we can discuss another time 😃 I'm not really in favour of this syntax

File: tests/cli/compress/conftest.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
+"""Fixtures for cli compress functions"""
+
+from typing import List
+
+import pytest
+
+
+class MockCompressAPI:
```
[the remark this answers is not in the document]
Comment (moonso) [reply tests/cli/compress/conftest.py:8]: Sometimes they are better to have close to where they are used. This is a big discussion and we need to formalise this much more. Atm I think the general apis like store and hk should be available from everywhere.

File: tests/cli/compress/test_cli_compress_fastq.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+def test_compress_fastq_cli_one_family(populated_compress_context, cli_runner, case_id, caplog):
+    """Test to run the compress command when no families are found"""
+    caplog.set_level(logging.DEBUG)
+    # GIVEN a context with a family
+
+    # WHEN running the compress command
+    res = cli_runner.invoke(fastq_cmd, [], obj=populated_compress_context)
+
+    # THEN assert the program exits since no cases where found
+    assert res.exit_code == 0
```
[the remark this answers is not in the document]
Comment (moonso) [reply tests/cli/compress/test_cli_compress_fastq.py:45]: Yes as long as it does not break

**Comment by moonso** [comment]
@barrystokman , since @patrikgrenfeldt is on vacation and we need a review from code owner I'm afraid you need to have a look at this...

**Comment by moonso** [comment]
@henrikstranneheim thanks for the review! Unfortunately there are some problems with GitHub and I can not use the add commit service... Guess I will try to add the changes manually 😁

**Comment by moonso** [comment]
Skärmavbild 2020-06-30 kl  14 30 00

Deployed!

File: cg/apps/crunchy/crunchy.py [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+
+    def spring_to_fastq(self, spring_path: Path):
+        """
+            Decompress SPRING into fastq by sending to sbatch SLURM
+        """
+        metadata_file = self.get_flag_path(spring_path)
+        spring_metadata = self.get_spring_metadata(metadata_file)
+        files_info = self.get_spring_archive_files(spring_metadata)
+
+        fastq_first_path = Path(files_info["fastq_first"]["path"])
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/apps/crunchy/crunchy.py:119]: Sounds intelligent. I think these are excellent discussion points for the future CG development meetings. I will start to sketch on an invitation and plan for those and will add this

File: cg/apps/crunchy/crunchy.py [hunk 14]
```diff
@@ -0,0 +1,486 @@
+"""
+    Module for compressing BAM to CRAM
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/apps/crunchy/crunchy.py:2]: I will resolve these as they are added on another PR

File: cg/apps/crunchy/crunchy.py [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+        if compression == "fastq":
+            return log_dir / "_".join([run_name, "compress_fastq.sh"])
+        if compression == "spring":
+            return log_dir / "_".join([run_name, "decompress_spring.sh"])
+        return log_dir / "compress_bam.sh"
+
+    @staticmethod
+    def get_file_type(file_path: Path) -> str:
+        """Check what file type a file is depending on the file ending"""
+        if str(file_path).endswith(SPRING_SUFFIX):
```
[the remark this answers is not in the document]
Comment (moonso) [reply cg/apps/crunchy/crunchy.py:329]: Good idea! Changed

**Pull request merged**